libpromises/eval_context.c

```diff
@@ -2111,6 +2116,17 @@ static Variable *VariableResolve(const EvalContext *ctx, const VarRef *ref)
         }
     }
 
+    // Try to qualify "this." variable to "current_bundle."
+    const Bundle *last_bundle = EvalContextStackCurrentBundle(ctx);
+    if (last_bundle && strcmp(ref->scope, SpecialScopeToString(SPECIAL_SCOPE_THIS)) == 0)
+    {
+        VarRef *qref = VarRefCopy(ref);
+        VarRefQualify(qref, last_bundle->ns, last_bundle->name);
+        var = VariableResolve(ctx, qref);
```

Comment:
This recursion messes me up. What is the guarantee that this will not recurse infinitely?

Reply:
I would say because `last_bundle->name` cannot be `this` so the scope will necessarily be different at the next call.